### PR TITLE
initial checkin of experimental Bareos REST API

### PR DESCRIPTION
**Pull request by aussendorf**
based on fastapi / uvicorn and python-bareos
  include systemtest with dependency on python3 modules fastapi and uvicorn

### CODE REVIEW

**Review by arogge — COMMENTED** [review]
* the `rest-api/main.py` is very large, maybe the model classes could be moved elsewhere or the functionality could be split up into separate files based on the subject
* there is a lot of unused configuration files in the system test. Usually we try to reduce the configuration to what is really required.
* if I counted correctly there are 65 routes and the systemtest tests only two of them, that's about 3% test coverage
* you're using `curl` in a shell-script for the test. As you're safe to assume that python is available, you might consider testing the API with a python program which will allow much more in-depth analysis of the results provided by the API
* some of the REST routes don't "feel" right, but maybe that's just me

File: rest-api/api.ini [hunk 1]
```diff
@@ -0,0 +1,11 @@
+[Director]
+Name=bareos-dir
+Address=127.0.0.1
+Port=9101
+
+[JWT]
+# to get a string like this run:
+# openssl rand -hex 32
+secret_key = 936959a2a6902056b924669796c74aad13b9da2b5cf637b70e377b3d7c29c6fb
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/api.ini:9]: it is not a random value, so we have to generate the key using openssl. I remove the default key and it will stop on startup, if no key is configured

File: rest-api/main.py [hunk 2]
```diff
@@ -0,0 +1,2595 @@
...
+
+# Read config from api.ini
+config = configparser.ConfigParser()
+config.read('api.ini')
+
+CONFIG_DIRECTOR_ADDRESS = config.get("Director","Address")
+CONFIG_DIRECTOR_NAME = config.get("Director","Name")
+CONFIG_DIRECTOR_PORT = config.getint("Director","Port")
+
+SECRET_KEY = config.get("JWT","secret_key")
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/main.py:46]: random doesn't work, I leave it blank and admins have to generate a key, see other comment below.

File: systemtests/tests/python-restapi/api/api.ini.in [hunk 3]
```diff
@@ -0,0 +1,11 @@
+[Director]
+Name=bareos-dir
+Address=@hostname@
+Port=@dir_port@
+
+[JWT]
+# to get a string like this run:
+# openssl rand -hex 32
+secret_key = 936959a2a6902056b924669796c74aad13b9da2b5cf637b70e377b3d7c29c6fb
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply systemtests/tests/python-restapi/api/api.ini.in:9]: we leave it blank, see other comment.

File: rest-api/api.ini (hunk 1 above)
[the remark this answers is not in the document]
Comment (arogge) [reply rest-api/api.ini:9]: Couldn't you just generate a value on startup if this isn't configured? This would be great for testing, because you don't need to configure it. In production - where you want a client's JWT to be valid after a restart - this would have to be configured, of course.

File: rest-api/main.py [hunk 4]
```diff
@@ -0,0 +1,2595 @@
...
+    except Exception as e:
+        raise HTTPException(status_code=500, detail="Could not read version from director")
+    if result and "version" in result:
+        return result["version"]
+    else:
+        response.status_code = 404
+        return {"message": "No version info returned"}
+
+
+@app.get("/control/directors/time", tags=["directors", "control"])
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/main.py:2503]: Taken from original specification.  All resource names are used in plural. Theoretically, the REST API can be used to connect to multiple directors.

File: systemtests/tests/python-restapi/api/startapi.sh.in [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+uvicorn main:app --port @restapi_port@ &
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply systemtests/tests/python-restapi/api/startapi.sh.in]: Done

File: rest-api/main.py [hunk 6]
```diff
@@ -0,0 +1,2595 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# BAREOS - Backup Archiving REcovery Open Sourced
+#
+# Copyright (C) 2014-2020 Bareos GmbH & Co. KG
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/main.py:5]: Done

File: rest-api/main.py [hunk 7]
```diff
@@ -0,0 +1,2595 @@
...
+CONFIG_DIRECTOR_PORT = config.getint("Director","Port")
+
+SECRET_KEY = config.get("JWT","secret_key")
+ALGORITHM = config.get("JWT","algorithm")
+ACCESS_TOKEN_EXPIRE_MINUTES = config.getint("JWT","access_token_expire_minutes")
+
+userDirectors = {}
+users_db = {}
+
+tags_metadata = [
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/main.py:53]: Done as yaml import

File: systemtests/tests/python-restapi/testrunner.in [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+start_bareos
+
+#run_python_unittests "$@"
+## place api tests here
+## TODO: error handling / make sure api gets stopped
+cd api
+./startapi.sh
+sleep 3
+echo "api started"
+TOKEN=`curl -X POST "http://127.0.0.1:@restapi_port@/token" -H  "accept: application/json" -H  "Content-Type: application/x-www-form-urlencoded" -d "grant_type=&username=admin-notls&password=secret&scope=&client_id=&client_secret=" | grep access_token | cut -d '"' -f 4`
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply systemtests/tests/python-restapi/testrunner.in:55]: find_program and condition added

File: rest-api/main.py [hunk 9]
```diff
@@ -0,0 +1,2595 @@
...
+            % (current_user.directorName, e)
+        }
+    if result and "time" in result:
+        return result["time"]
+    else:
+        response.status_code = 404
+        return {"message": "No time info returned"}
+
+
+@app.put("/control/directors/reload", tags=["directors", "control"])
```
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/main.py:1803]: duplicate, see above

File: rest-api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (joergsteffens) [reply rest-api/main.py:5]: No, not yet done.

File: docs/pkglists/CentOS_7.x86_64 [hunk 10]
```diff
@@ -1,10 +1,8 @@
 debug/x86_64/bareos-debuginfo
```
Comment (joergsteffens) [inline docs/pkglists/CentOS_7.x86_64:1]: This file should not be part of this commit. I guess, it will disappear with a rebase from master.

File: rest-api/main.py [hunk 11]
```diff
@@ -0,0 +1,1824 @@
+#!/usr/bin/env python
```
Comment (joergsteffens) [inline rest-api/main.py:1]: Is it an option to rename this file from main.py to something that indicate its purpose? Like bareos-rest-api.py. Otherwise it may confusing in the process list when running.

File: rest-api/main.py [hunk 12]
```diff
@@ -0,0 +1,1824 @@
...
+    jobDef: jobResource = Body(..., title="Job resource"),
+    response: Response,
+    current_user: User = Depends(get_current_user),
+):
+    """
+    Create a new job resource.
+    Console command used: _configure add job_
+    
+    TODO: string lists are processed and acknowledged in director's response
+    but do not make it into the configuration.
```
Comment (joergsteffens) [inline rest-api/main.py:1026]: This is fixed by now.

File: rest-api/requirements.txt [hunk 13]
```diff
@@ -0,0 +1,12 @@
+dataclasses==0.7
```
Comment (joergsteffens) [inline rest-api/requirements.txt:1]: Are fixed version really required? Especially for python-bareos, which will be shipped together with the rest-api?

File: rest-api/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (aussendorf) [reply rest-api/main.py]: Now donw

File: rest-api/main.py (hunk 11 above)
Comment (aussendorf) [reply rest-api/main.py]: renamed to bareos-restapi.py

**Comment by fbergkemper** [comment]
The README of this tech preview should be adapted to also address end users who want to try it out.

- Python module requirements
- DIRD User/Console/Profile resource requirements

File: rest-api/README.md [hunk 14]
```diff
@@ -0,0 +1,65 @@
+# Bareos REST API using fastAPI and python-bareos.
+
+Experimental and subject to enhancements and changes.
+
+This is an experimental backend for development purposes.
+
+It provides a REST API using fastapi and python-bareos to connect to
+a Bareos director.
+
+Connexion handles the HTTP requests defined by using OpenAPI
```
Comment (fbergkemper) [inline rest-api/README.md:10]: This is no longer the case and relates to a previous approach.

File: rest-api/README.md [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+specification.
+
+* https://fastapi.tiangolo.com/
+* https://connexion.readthedocs.io/en/latest/
```
Comment (fbergkemper) [inline rest-api/README.md:14]: see above

File: rest-api/requirements.txt (hunk 13 above)
Comment (fbergkemper) [reply rest-api/requirements.txt:1]: It is not required but basically a good idea to pip freeze dependency installed in your virtual environment to avoid trouble and headache. Some versions of packages will not work well with other packages and their respective versions.

Comment (fbergkemper) [reply rest-api/requirements.txt:1]: For the package which is pip installed, dependencies should not rely on the requirements.txtbe and instead be declared like described in https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html?highlight=declaring%20dependencies#declaring-required-dependency .

File: rest-api/requirements.txt [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+pydantic==1.6.1
+python-bareos==19.2.8
+# This will require gcc and some headerfiles
+#sslpsk==1.0.0
+starlette==0.13.6
+python-jose==3.2.0
+uvicorn==0.11.8
+python-multipart==0.0.5
+passlib==1.7.2
+packaging==20.4
```
Comment (fbergkemper) [inline rest-api/requirements.txt:12]: pyyaml is missing

ModuleNotFoundError: No module named 'yaml'

File: rest-api/requirements.txt (hunk 13 above)
Comment (aussendorf) [reply rest-api/requirements.txt]: requirements.txt adapted, only starlette needs a max version number. Packaging of rest-api for pip will be next step on the list.

File: rest-api/requirements.txt (hunk 16 above)
Comment (aussendorf) [reply rest-api/requirements.txt]: pyyaml now part of requirements.

File: systemtests/tests/python-restapi/api/stopapi.sh [hunk 17]
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+PID=`cat api.pid`
+kill $PID
+rm api.pid
```
Comment (fbergkemper) [inline systemtests/tests/python-restapi/api/stopapi.sh:4]: EOF newline missing

File: rest-api/README.md (hunk 15 above)
Comment (aussendorf) [reply rest-api/README.md]: removed connexion remark

File: rest-api/README.md (hunk 14 above)
Comment (aussendorf) [reply rest-api/README.md]: removed connexion remark

File: rest-api/main.py (hunk 12 above)
Comment (aussendorf) [reply rest-api/main.py]: comment removed

File: systemtests/tests/python-restapi/api/stopapi.sh (hunk 17 above)
Comment (aussendorf) [reply systemtests/tests/python-restapi/api/stopapi.sh]: fixed

File: rest-api/bareos-restapi.py [hunk 18]
```diff
@@ -0,0 +1,1796 @@
...
+
+@app.put("/control/directors/reload", tags=["directors", "control"])
+def read_director_time(
+    *, response: Response, current_user: User = Depends(get_current_user)
+):
+    """
+    Reload director configuration from files. Command used: _reload_
+    """
+    result = None
+    listCommand = "reload"
```
Comment (fbergkemper) [inline rest-api/bareos-restapi.py:1783]:
```suggestion
    updateCommand = "reload"
```

File: rest-api/bareos-restapi.py [hunk 19]
```diff
@@ -0,0 +1,1796 @@
...
+
+@app.get("/control/directors/version", tags=["directors", "control"])
+def read_director_version(
+    *, response: Response, current_user: User = Depends(get_current_user)
+):
+    """
+    Read director version. Command used: _version_
+    """
+    result = None
+    listCommand = "version"
```
Comment (fbergkemper) [inline rest-api/bareos-restapi.py:1739]: I'm not sure if this should be named listCommand. It's a read operation but a general DIR command is used instead of the (l)list command.

File: rest-api/bareos-restapi.py [hunk 20]
```diff
@@ -0,0 +1,1796 @@
...
+
+@app.get("/control/directors/time", tags=["directors", "control"])
+def read_director_time(
+    *, response: Response, current_user: User = Depends(get_current_user)
+):
+    """
+    Read director time. Command used: _time_
+    """
+    result = None
+    listCommand = "time"
```
Comment (fbergkemper) [inline rest-api/bareos-restapi.py:1759]: See above.

Comment (aussendorf) [reply rest-api/bareos-restapi.py:1759]: done

File: rest-api/bareos-restapi.py (hunk 19 above)
Comment (aussendorf) [reply rest-api/bareos-restapi.py:1739]: done

File: rest-api/bareos-restapi.py (hunk 18 above)
Comment (aussendorf) [reply rest-api/bareos-restapi.py]: done

File: debian/control [hunk 21]
```diff
@@ -349,6 +349,8 @@ Description: Backup Archiving Recovery Open Sourced - common tools
 
 
 
+
```
Comment (fbergkemper) [inline debian/control:352]: I guess this is an accident. The debian/control file should not be a part of this PR, please remove.

**Review by fbergkemper — APPROVED** [review]
Looks good to me so far as a tech preview.

**Review by franku — CHANGES_REQUESTED** [review]
Huge amount of work! However I have some smaller issues: 
- Please update the hyperlinks to documentation as suggested (some have errors)
- the build environment is currently located in the source tree but should be located in the build area

**Review by franku — APPROVED** [review]
